Source/x11/XGServerWindow.m

```diff
@@ -4430,74 +4436,154 @@ - (void) freecursor: (void*) cid
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSArray        *screens;
+  NSMutableArray *tmpScreens;
+  int            i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+    monitors = NULL;
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  RROutput            primary_output;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+  int                 mi;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  if (screen_res != NULL)
+    {
+      monitorsCount = screen_res->noutput;
+      tmpScreens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      primary_output = XRRGetOutputPrimary(dpy, [self xDisplayRootWindow]);
+      for (i = 0, mi = 0; i < screen_res->noutput; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+              
+              monitors[mi].screen_id = defScreen;
+              monitors[mi].depth = [self windowDepthForScreen: mi];
+              monitors[mi].resolution = [self resolutionForScreen: defScreen];
+              monitors[mi].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                              crtc_info->width, crtc_info->height);
+              /* Add monitor ID (index in monitors array).
+                 Put primary monitor ID index 0 since NSScreen get this as main 
+                 screen if application has no key window. */
+              if (screen_res->outputs[i] == primary_output)
+                {
+                  [tmpScreens insertObject: [NSNumber numberWithInt: mi] atIndex: 0];
+                }
+              else
+                {
+                  [tmpScreens addObject: [NSNumber numberWithInt: mi]];
+                }
+              mi++;
+            }
+        }
+      XRRFreeScreenResources(screen_res);
+      screens = [NSArray arrayWithArray: tmpScreens];
+    }
+#endif
+
+  if (monitorsCount == 0)
+    {
+      /* It is assumed that there is always only one screen per application. 
+         We only need to know its number and it was saved in _initXContext as
+         `defScreen`. */
+      monitorsCount = 1;
+      monitors = NSZoneMalloc([self zone], sizeof(MonitorDevice));
+      monitors[0].screen_id = defScreen;
+      monitors[0].depth = [self windowDepthForScreen: defScreen];
```

Comment:
Does this method get called with a monitor or a screen id?

Reply:
Monitor. In this case it should be `0`. Fixed.